log.go

```diff
@@ -0,0 +1,48 @@
+package marker
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/fatih/color"
+)
+
+type LogMarkerOption func(*LogMarker)
+
+type LogMarkerRule struct {
+	Matcher MatcherFunc
+	Color   *color.Color
+}
+
+type LogMarker struct {
+	rules  []LogMarkerRule
+	logger *log.Logger
+}
+
+func NewLogMarker(options ...LogMarkerOption) *LogMarker {
+	logMarker := &LogMarker{logger: log.New(os.Stdout, "", 0)}
+	for _, option := range options {
+		option(logMarker)
+	}
+	return logMarker
+}
+
+func SetLogger(logger *log.Logger) LogMarkerOption {
+	return func(logMarker *LogMarker) {
+		logMarker.logger = logger
+	}
+}
+
+func (l *LogMarker) AddRule(rule LogMarkerRule) *LogMarker {
+	l.rules = append(l.rules, rule)
+	return l
+}
+
+func (l *LogMarker) Print(v ...interface{}) {
+	marked := fmt.Sprint(v...)
+	for _, rule := range l.rules {
```

Comment:
How do you feel about the "constructor" for the `LogMarker` taking in a `MarkBuilder` instead? that way there's a single piece of the application that is always responsible for marking text, and you can avoid the logger having to maintain any state with regards to marking rules.

Reply:
`MarkBuilder` marks the text immediately and returns the marked text when finally `Build` called, so it doesn't store the mathcer functions or colors. I couldn't think the usage of `MarkBuilder` for applying same rules over log stream.

`Mark` function in `marker.go` file is the only responsible for marking text anyway.

Therefore, even if we use `MarkBuilder` for marking (we are just looping over rules and applying them instead of using `MarkBuilder`), we must store the marking rules specified by user to apply them for every log.

Lets also think about the UX of the log marker API, in my opinion we need to provide a lean interface with minimum amount of boilerplate code to the user. Can you give me an example usage with `MarkBuilder` from the user perspective("constructor" parameters etc.) for me to understand better, like (current implementation usage):

```golang
red := color.New(color.FgRed)
blue := color.New(color.FgBlue)
// function names most probably will change, take it as pseudo :P
logger := marker.NewLogger() 
logger.AddRule(marker.Rule{marker.MatchBracketSurrounded(), red}).
         AddRule(marker.Rule{marker.MatchParensSurrounded(), blue}) // just stores the rules
// Implementing AddRules([]marker.Rule) function can be beneficial to reduce struct literal boilerplate 
logger.Print("[Error] we have errors and we can see them!") // applies the stored rules before sending the log to the downstream log.Print function
logger.Print("(WOW) marker is [awesome] !") // applies to this too!
```

Correct me if I understood you wrong.

Reply:
You understood me correctly! I had a misunderstanding about the way the builder operated. Makes perfect sense now :+1:. In suggesting the `MarkBuilder` I was trying to find a way where the actual logic for marking the strings lives in a central place and the logger is only responsible for facilitation. In thinking over it some more, the `Mark` and `MatchFunc` methods take care of this concern nicely so no need to over-engineer it 👍 

Approach makes sense, and thanks for the explanation!